aten/src/ATen/native/TensorConversions.cpp

```diff
@@ -1399,90 +1415,90 @@ TORCH_IMPL_FUNC(_convert_indices_from_csr_to_coo_structured_cpu)
  * https://github.com/scipy/scipy/blob/8a64c938ddf1ae4c02a08d2c5e38daeb8d061d38/scipy/sparse/sparsetools/csr.h
  * Modified to ensure sorted BSR column indices.
  */
-template <class I, class T>
-void _csr_to_block_csr_cpu_kernel(
-    const I n_row,
-    const I n_col,
-    const I R,
+template <class I, class T, bool compressed_rows>
+void _compressed_to_block_compressed_cpu_kernel(
+    const I n_dim_compressed,
+    const I n_dim_plain,
     const I C,
-    const I* input_crow_indices,
-    const I* input_col_indices,
+    const I P,
+    const I D,
+    const I* input_compressed_indices,
+    const I* input_plain_indices,
     const T* input_values,
-    I* result_crow_indices,
-    I* result_col_indices,
+    I* result_compressed_indices,
+    I* result_plain_indices,
     T* result_values) {
-  // All blocks are possible, that is, may be allocated if a single non-zero
-  // value lives within them. Otherwise they're not.
-
-  // Allocate pointers for all possible column blocks plus 1
-  std::vector<T*> blocks(n_col / C + 1, (T*)0);
+  // All blocks are possible, that is, may be allocated if a single
+  // non-zero value lives within them. Otherwise they're not.
```

Comment:
Is it tested? Looks like there was a single line that was supposed to error out, and now it is removed, so, I guess, only rectangular blocks are being checked.

Reply:
Previously, only square blocks were supported, this PR extends this to rectangular blocks.
There are no other blocks than rectangular blocks, so, can you be more specific about "only rectangular blocks"?

Reply:
Sorry, I mean square blocks are only tested, not generic rectangular.

Reply:
Indeed, TestSparseAny::test_to_sparse tests CSC->BSC conversion using blocksize == (1, 1) only. @alexsamardzic could you apply the following patch:
```
diff --git a/test/test_sparse.py b/test/test_sparse.py
index be77420fac..4355e489f3 100644
--- a/test/test_sparse.py
+++ b/test/test_sparse.py
@@ -4421,6 +4421,21 @@ class TestSparseAny(TestCase):
                 r2 = explicit_to_sparse(t)
                 self.assertEqual(r2, r)
 
+                # Check inverse conversion from sparse compressed block tensors
+                if from_layout == torch.sparse_bsr:
+                    batch_ndim = t.crow_indices().dim() - 1
+                    from_blocksize = t.values().shape[batch_ndim + 1:batch_ndim + 3]
+                elif from_layout == torch.sparse_bsc:
+                    batch_ndim = t.ccol_indices().dim() - 1
+                    from_blocksize = t.values().shape[batch_ndim + 1:batch_ndim + 3]
+                else:
+                    continue
+                if r.ndim != 2:
+                    continue
+
+                t2 = r.to_sparse(layout=from_layout, blocksize=from_blocksize)
+                self.assertEqual(t2, t)
+
         # extra tests
         if (from_layout, to_layout) == (torch.sparse_csr, torch.sparse_bsr):
             # See gh-90910
```
that will enable non-square block conversion tests.

Reply:
Thanks - applied, and the tests pass.

Reply:
Btw, `if r.ndim != 2:` was a quick way to skip the test for unsupported batch dimensions but it may be too restrictive. For instance, hybrid cases are not tested.

Reply:
Better safe than sorry, this one guarantees rectangular shapes, right?